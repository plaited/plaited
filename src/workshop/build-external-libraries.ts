import { zip } from './workshop.utils'

// Define the base artifact type from Bun
type BunBuildArtifact = Bun.BuildOutput['outputs'][number]

// Define a more specific type for entry point artifacts using an intersection
// This ensures it has all properties of BunBuildArtifact AND the specific overrides
type EntryPointBuildArtifact = BunBuildArtifact & {
  kind: 'entry-point' // Narrows the kind
  entrypoint: string // Ensures entrypoint is a non-optional string
}

const isEntryPoint = (output: BunBuildArtifact): output is EntryPointBuildArtifact => output.kind === 'entry-point'

export const buildExternalLibraries = async (
  externalLibraries: Set<string>,
  responseMap: Map<string, Response>,
): Promise<Record<string, string>> => {
  const importMap: Record<string, string> = {}
  const buildEntrypoints: string[] = []
  const resolvedPathToLibraryName: Record<string, string> = {}
  const libraries = new Set(['plaited/jsx-runtime', ...externalLibraries])

  for (const libraryName of libraries) {
    try {
      const resolvedPath = Bun.resolveSync(libraryName, process.cwd())
      buildEntrypoints.push(resolvedPath)
      resolvedPathToLibraryName[resolvedPath] = libraryName
    } catch (error) {
      console.warn(`Could not resolve external library "${libraryName}": ${(error as Error).message}`)
    }
  }

  if (buildEntrypoints.length === 0) {
    // This means all libraries failed to resolve or the input array was effectively empty
    return {}
  }

  const { outputs } = await Bun.build({
    entrypoints: buildEntrypoints,
    splitting: true, // Recommended for potentially shared dependencies among external libs
    naming: 'node_modules/[name].[ext]', // Outputs to _node_modules directory structure
    sourcemap: 'inline',
  })
  await Promise.all(
    outputs.map(async (output) => {
      // We are interested in the entry points to map them back to the original library name
      if (isEntryPoint(output)) {
        // Cast to the more specific type after checking the kind.
        const entryPointArtifact = output
        const originalLibraryName = resolvedPathToLibraryName[entryPointArtifact.entrypoint]
        if (originalLibraryName) {
          // Ensure the path starts with a '/' for typical import map usage
          const path = entryPointArtifact.path.startsWith('/') ? entryPointArtifact.path : `/${entryPointArtifact.path}`
          importMap[originalLibraryName] = path
          const code = await output.text()
          responseMap.set(path, zip(code))
        }
      }
    }),
  )
  return importMap
}
