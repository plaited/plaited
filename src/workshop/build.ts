import { usePlugin } from './use-plugin.js'

export const build = async (root: string, entrypoints: string[]) => {
  const isProd = process.env.NODE_ENV === 'production'
  return await Bun.build({
    root,
    entrypoints,
    sourcemap: isProd ? 'none' : 'inline',
    minify: isProd,
    splitting: true,
    publicPath: '/',
    plugins: [usePlugin(root)],
  })
}
