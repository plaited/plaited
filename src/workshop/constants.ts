export const defaultRegistry = 'registry.ts'
export const defaultTestExtension = '.playwright.spec.ts'
export const playwrightVersion= '1.30.0'
export const yarnVersion = '3.3.1'