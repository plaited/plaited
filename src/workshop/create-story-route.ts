import path from 'node:path'
import { STORIES_FILTERS_REGEX } from 'plaited/testing'

import { kebabCase } from 'plaited/utils'

/**
 * Creates a standardized route path for story files in the workshop system.
 * Converts file paths and export names to kebab-case for consistent URL structure.
 *
 * @param options - Configuration object
 * @param options.relativePath - The relative file path of the story
 * @param options.exportName - The exported story name
 * @returns A formatted route path string in the format "{directory}/{basename}--{storyName}"
 *
 * @example
 * ```ts
 * // For a file "/src/components/Button.stories.tsx" with export "PrimaryButton"
 * createStoryRoute({
 *   filePath: "/src/components/Button.stories.tsx",
 *   exportName: "PrimaryButton"
 * });
 * // Returns: "/src/components/button--primary-button"
 * ```
 */
export const createStoryRoute = ({ filePath, exportName }: { filePath: string; exportName: string }) => {
  const dirname = path.dirname(filePath)
  const basename = kebabCase(path.basename(filePath.replace(STORIES_FILTERS_REGEX, '')))
  const storyName = kebabCase(exportName)
  const id = `${basename}--${storyName}`
  return `${dirname}/${id}`
}
