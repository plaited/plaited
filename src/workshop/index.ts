export { getRoutes, ssr } from './utils/index.ts'
export * from './types.ts'
export * from './build.ts'
export * from './start.ts'
