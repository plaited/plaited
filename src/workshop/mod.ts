export { livereloadTemplate } from '../server/mod.ts'
export { type StoryConfig, type Story, type WorkshopConfig } from './types.ts'
export * from './start.ts'
export * from './setup.ts'