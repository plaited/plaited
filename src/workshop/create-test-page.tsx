import path from 'node:path'
import { type StoryObj } from '../testing/assert.types.js'
import { PlaitedFixture } from '../testing/plaited-fixture.js'
import { ssr } from '../jsx/ssr.js'
import { PLAY_EVENT, PLAITED_FIXTURE } from '../testing/assert.constants'
import { LIVE_RELOAD_PATHNAME, cacheBustHeaders } from './workshop.utils.js'

export type Createstpage = {
  story: StoryObj
  route: string
  responses: Map<string, Response>
  filePath: string
  exportName: string
  port: number
}

export const getLiveReloadScript = ({
  port,
  route,
  filePath,
  entryPath,
  exportName,
}: {
  port: number
  route: string
  filePath: string
  entryPath: string
  exportName: string
}) => `
const eventSource = new EventSource("http://localhost:${port}${LIVE_RELOAD_PATHNAME}");

eventSource.addEventListener('open', async (event) =>{
  console.log("SSE Connection Opened:", event);
  const { ${exportName} } = await import('${entryPath}');
  await customElements.whenDefined("${PLAITED_FIXTURE}")
  const fixture = document.querySelector("${PLAITED_FIXTURE}");
  fixture.trigger({
    type: '${PLAY_EVENT}',
    detail: {
      route: "${route}",
      filePath: "${filePath}",
      entryPath: "${entryPath}",
      exportName: "${exportName}",
      story: ${exportName}
    }
  });
});

eventSource.addEventListener('reload', (event) => {
    console.log("SSE 'reload' Event Received. Data:", { reloads: event.data });
    console.log("Reloading page in a moment...");
    // Reload the page after a short delay
    setTimeout(() => {
        window.location.reload();
    }, 750);
});

eventSource.addEventListener('error', (event) => {
    console.error("SSE Error Occurred:", event);
    if (eventSource.readyState === EventSource.CLOSED) {
        console.warn("SSE Connection was closed.");
    } else if (eventSource.readyState === EventSource.CONNECTING) {
        console.warn("SSE Connection is trying to reconnect.");
    }
})
`

export const createTestPage = ({ story, route, responses, filePath, exportName, port }: Createstpage) => {
  const entryPath = filePath.replace(/\.tsx?$/, '.js')
  const args = story?.args ?? {}
  const styles = story?.parameters?.styles ?? {}
  const headers = story?.parameters?.headers?.(process.env) ?? new Headers()
  const tpl = story?.template
  const page = ssr(
    <html>
      <head>
        <title>Story:{path.basename(route)}</title>
        <link
          rel='shortcut icon'
          href='#'
        />
      </head>
      <body>
        <PlaitedFixture
          children={tpl?.(args)}
          {...styles}
        />
        <script
          trusted
          type='module'
        >
          {getLiveReloadScript({ port, exportName, filePath, entryPath, route })}
        </script>
      </body>
    </html>,
  )
  responses.set(
    route,
    new Response(`<!DOCTYPE html>\n${page}`, {
      headers: {
        ...cacheBustHeaders,
        ...headers,
      },
    }),
  )
}
