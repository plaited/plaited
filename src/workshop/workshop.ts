import { globEntries, buildEntries } from './workshop.utils.js'
import { type StoryObj, type TestParams } from '../testing/assert.types.js'
import { createStoryRoute } from './create-story-route.js'
import { createTestPage } from './create-test-page.js'
import type { WorkshopParams } from './workshop.types.js'

export const workshop = async ({ cwd, output, background, color, designTokens }: WorkshopParams) => {
  const stories = new Map<string, TestParams>()
  const responses = new Map<string, Response>()
  const entrypoints = await globEntries(cwd)
  const values = await Promise.allSettled(
    entrypoints.flatMap(async (entry) => {
      const { default: _, ...rest } = (await import(entry)) as {
        [key: string]: StoryObj
      }
      return await Promise.all(
        Object.entries(rest).map(async ([exportName, story]) => {
          const filePath = entry.replace(new RegExp(`^${cwd}`), '')
          const route = createStoryRoute({ filePath, exportName })
          return await createTestPage({
            output,
            story,
            route,
            entry,
            exportName,
            background,
            color,
            designTokens,
          })
        }),
      )
    }),
  )
  const htmlEntries = values.flatMap((settled) => (settled.status === 'fulfilled' ? settled.value : []))
  await buildEntries({ output, htmlEntries, responses })
  return stories
}
