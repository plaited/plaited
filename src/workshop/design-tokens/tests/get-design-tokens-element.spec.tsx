import { test, expect } from 'bun:test'
import { getDesignTokensElement } from 'plaited/workshop'
import beautify from 'beautify'

test('getDesignTokensElement', () => {
  let DesignTokens = getDesignTokensElement(`:host{--blue: blue}`)
  expect(beautify((<DesignTokens />).html.join(''), { format: 'html' })).toMatchSnapshot()
  DesignTokens = getDesignTokensElement(`:host{--blue: blue}`, 'plaited-tokens')
  expect(
    beautify(
      (
        <DesignTokens>
          <div>hello world!</div>
        </DesignTokens>
      ).html.join(''),
      { format: 'html' },
    ),
  ).toMatchSnapshot()
})
