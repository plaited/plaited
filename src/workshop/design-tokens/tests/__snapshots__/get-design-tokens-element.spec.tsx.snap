// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`getDesignTokensElement 1`] = `
"<design-tokens>
    <template shadowrootmode="open" shadowrootdelegatesfocus>
        <style>
            :host{display:contents;}:host{--blue: blue}
        </style>
        <slot></slot>
    </template>
</design-tokens>"
`;

exports[`getDesignTokensElement 2`] = `
"<plaited-tokens>
    <template shadowrootmode="open" shadowrootdelegatesfocus>
        <style>
            :host{display:contents;}:host{--blue: blue}
        </style>
        <slot></slot>
    </template>
    <div>hello world!</div>
</plaited-tokens>"
`;
